### ISSUE TITLE
GetExternalLoginInfoAsync is null when called from ExternalLoginModel.OnGetCallbackAsync - IFrames involved.

### ISSUE THREAD

**Issue by ghstahl** [dotnet/aspnetcore#25375]
Goal: Perform a silent login using a hidden iframe to google.
Result: I can achieve this by passing prompt=none to google which then doesn't x-frame-options=deny me.
All Good when I run the app normally and I route to ``` https://localhost:5001/Plugin ``` and ``` https://127.0.0.1.xip.io:5001/Plugin ```  

My Plugin Page has my hidden iFrame.  

```
<button id='signin'>signin</button>
<p id="received-message">I've heard nothing yet</p>
<form style="display: none;" id="loginform" target="guestFrame"
      
      action="/Identity/Account/ExternalLogin?returnUrl=%2FOIDCIFrameResult&amp;errorUrl=%2FOIDCIFrameResult&amp;prompt=none&amp;provider=google"
      method="post">
    <div>
        <p>
             
            <input type="submit">
        </p>
    </div>
</form>
<iframe style="display: none;" id="guestFrame" name="guestFrame" src="/Blank">
    Your browser does not support inline frames.
</iframe>
```

Ok, now put the entire app in an iFrame and try to do a silent login under that context.
```
<iframe id="guestFrame" name="guestFrame" src="https://127.0.0.1.xip.io:5001/Plugin">
</iframe>
```
The above ```iFrame``` is hosted in ```https://localhost:5001/IFrameHost```  
I am using [xip.io](https://xip.io) as my wildcard DNS.  ```https://127.0.0.1.xip.io:5001```
My ```signin-google``` gets hit and it looks like there is a good code in there.
```
GENERAL
Request URL: https://127.0.0.1.xip.io:5001/signin-google
Request Method: POST
Status Code: 302 
Remote Address: 127.0.0.1:5001
Referrer Policy: no-referrer-when-downgrade

Response Headers
cache-control: no-cache
content-length: 0
date: Fri, 28 Aug 2020 20:51:15 GMT
expires: Thu, 01 Jan 1970 00:00:00 GMT
location: /Identity/Account/ExternalLogin?returnUrl=%2FOIDCIFrameResult&handler=Callback
pragma: no-cache
server: Kestrel
set-cookie: .AspNetCore.Correlation.google.gxGuRZ5ZVbOr4SmC3zd6CLQpHZ_j-ZTc-jyvSnAF-IQ=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/signin-google; secure; samesite=none; httponly
set-cookie: .AspNetCore.OpenIdConnect.Nonce.CfDJ8L16rcoVAjJMt5TusKivrFcSEujxcxPIbAwED3uPgd2mk-J_fEJKkogtr3NaSVolWh5Rv5DcehuReLphIyvY7VG-iypvHqGc7knVZXv2FWJaFed1bQaA2jkGFbmEmQs4DydnmwOdlA4Df_o8SjCPEW8Bq54SWJz6uNI8L9p0_e8ZllW1gaF0zCOuoQn_XdEm8jUW6o1h2N1o9YLFAbj3-w0rj5BWqOfl6ysEkeY0DXe7xm8Ycmd7PpC98edYtnkUgs5Kfs7H_4bn0E4TSP_1XS8=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/signin-google; secure; samesite=none; httponly
set-cookie: Identity.External=CfDJ8L16rcoVAjJMt5TusKivrFecjgIwf31r3SrfRXOwnrnUe9wLJ38mJtd8HAzKOc_2rsgrFsLc6lAnKcHa5pYmafjUFaKJcmBZt2ys2JSUl3qFKVqSTZaSpS4E21Tt70yaz2T9AzSzDOhBv2hIBrtj1Y9c1Ximra9zPlTXLI1PuUNPl7ZQSTzQfep7-hhr7Fj0aCz4Qq14UvjHBaU8QjGd_XACdODJc4aX1PlT8_JgXDl5ob9RtDxmta1egpT6EkFOFwBC1Ngb0UiY5Ilz-c1R3255SDIxa6PVVNjy55ccFo0Hpfik9JWj6oTgGApt6koNrxbFOTK5sLAt5S7lR1FbL8Jkq_DOpmTzfSgEPVV7PIfW6AK4zZav6qawqgTn01mYRM-NO8f_gcrhTBfQpEbrYAFqVTu-Svo6ZkKTI-iRMe57arPz8e8nUvRnqplAvR8MbRKvXNAHGdRcN-JDvzxZQHxm7I0u-gX0MUm1CezYDhbaENhosyZrcxZhNZ9u7XdVgEtAlPe_KX1Vdn0zHmgHadcW_q8l1yvVzMTigryWr8-6wvGyeCPVhGGfRH6vIqkF9hGgTNXLpIso3xESmRJveCXbYWZfZ9Tm2w51T4ToCwOc5ymNmYkUBvQYX0HCpKkwOIlCn1JLGWy7q6WcN1VRcZHImvX-TA3R9aiIqRwbuAe3Ju4Qop8oAGoXW1FxFPAGIQRytj_m7uPBdngitb5bZgcFKhXr5vgxngWrE2Ns55ItXu85aGHi8DymfUmuH-Wtsd6haLpq39E08D_zudyY11hbt2_raO9OPNkWV8R_QCYR4eij7sp1OTG_Jq1tkdsQRTe0DfmxE9-reYYw8hK_sF4Hb8GU8QO0Zrn0Ia9UrF6tqW-LNhO5Id02Jmxl2Yn6yOipNwO9FbRDmLy9O4LumZWZw1miLS-XJ9KxPRPcR3x_wCONtKrxfWZzEqYs7T34CSQB71okiKFwfMu_vUNJl44TUsmgZ9QEtgIxlHcSeIHPc7Y2FYU3ZEM3gBmekbGGywzQ48TxNedsHo9PJ_We5BHbyBQK5XvVJ8r22VBOYlZHaNbcOt3WS7KxJp-uHmlfCY7NuC5d27IqVN0Ir5sLDbnr6KSblI2y82faQXoXUQz-HQlTly4O6lEJc1YbEAcfgHWieye-3A9wbjEt6N7HISTjcWWNgJIk_IfJcQC2lP95wP5VHZgwKQrIitkRCYaUxwfTEqUzVZq34chEd9Y8eF1A-LAJjHpT4nsovrRuD1qof63XmJ-dqwdAEeC41CLRDgWRZRdAsFj2xEZkTIDu8IQZugOfxucgYBH8TxWIssU7YLtR4CVQoDOCmT-EIodwEPAY0Bi_mMHh6O0EhrGyXsvEFFm_M-A-6uLZgrL8jjU2DNUG9ZhFSIUncDZ1_9iEvuZfv8OMUCWayEmmml8-SXeElPvZTDk4O2Els11JRP7q1jTHZ-5vakdbRDql68HiHryY8xEsDN_4LW_blaHCwKV32npzrB_vnETmj0Ii7M1QytZV3Nfdr8R-bfYg6wF46z5bm-9TwWkX4_pwJ5dp116kvE5LkADo-NIrNo33EJCsp-iYYhmUWRVL5Rh7nszVHr-1fj2TramHcKRn7_oi4djcFIHl0w-i4dF3kRPy-dgE2nKHsSM9Rpyjn-whNSQfPM7AOxIYTbllejAmGVJxUU83TIrQiOeNxml6e5mlp3GrVmC5qtBmnUA4hdrpyGgWFzXx5uBCtLN2z39HrrKvIKD25naRPqu6mRHI5QRqSFL9mpmv4Cseqk975MnuUPtjI-8BhpGoavN_4PALb_VeM-JTh3TW0W40S6BOi3c97ZojY_J34EEO-WSk5zNw1WHS233etkQ5qCsjtae2f3_dC9Gi_SI9CgFgEnz6LjqqaXOisgVGJTGC-D4V-Ms0teHoEklFYrvhUMZiCmxGdW8FwrZF6bG9uAKX1iRvg_557SIczHx24v9zXZL33dN0m11THnnD_0yVAc6ac9uiXBWmI0eZ-0jxk0Rg2dZsZ75RS-Mj8CNQZPfPfvfQj9LpZh8uyS3c3IPJVLzLyVpmEC0co-8gty45-qFU-cjoyiPI3v9hH9HWcO6bcdp7w1dgj-nQxPCj0Sb8MkPbW2cvl2y9Hwr5MmAjEp8gnCtLlfvm7hbbzwntkQURVP2V7ir83ZjNfObKcXN7SPbcsf88yTNfnilrLyF-GfaufOiqz5PEEy33k2geQMRbjbjdzpEarQjLZnQGLZ5hupMmbr7UdME2e1PCZPBp-KWZP4qA8uW3gCsmSqhBQgo4fcmjFocPPyt1IEnhmJqfI8c32_CvCK8Df98f183xwuBBVJr23MN1rVYf4avXEnrTfrs6I6CJfy8ZN5upxcUqpMxTQiMcO1lVycx8WYvcZTl1VXc9CipLiE7t92rqLeoyUQENWSvryTIL5JDtEexJTZ5fGDyLc1NCq5JidkG-oH32aPtSwm1hEQa8MsWfK64N_8f5YA5KVQuen-d-IQil_u7LVNS5BfIjEuhydZqRdoEi-8IKoraDCFzs6prdaXzqB3rQJwa2cOnuCjQNT8xz2kvRBCq75rt7I-srPSYNA57W9SwP5dn8GR2AxUI8CcO7zCfuBwO4FbcEgBtxCWaDmhAVnZd91MH0sslMo4hqNI8PQnNNvtGDXuHsM68tETZflcEW1EEH6Ylr2LWCDGlEZEK6PI1QyYI8TteMw0Rv6dxHbJqhYtvmb3Mx9FIf48NCEWNOnHK6DLt4R6gf-DjZp8pe8z1yMQ8sUccE_4R53irREwFrwij8BUHm1ncDzAVBK8uaGoHskM4oeLjKn8xCEW5pOuOBzXryrt9d0-3ymmLXUsI8du2WB0LmsMOJ3IM4f7zi9P-TjWTt4bdJ7YAmF5lp_zU0ltvMexKB9aTj-b7gfoEd2nBL3rd4E_VCqh_WosaI_t27CulyuuWCsGEFChO1vC0ONeyKHcjJSBv7E3kUYoYvo664SGC8OBZe69VCxFSPOhersp8XS74F6VtTzj7B8xvht8IsuOmZ7Ce5dj6ZuacUarAVppJBzeFfE2HrDOKLQGZ9Irq9berzA9zmQlHFDGNbj3GkZ5x8RmAc-pn1w_GAHeIVzULfNZ8Gxc-3AkqwlqkbDVZEuUL10zgtjh9vezwhiAvllcKS86AunCdb6UKegoSKPHmAWHQBLNwVYBjs5lDBOlxpzN41UhHAAlKmBfBS4QDoejTZN8yZGWS7irf-9VmBDiI1oSh6UbajAXLVmKJ6cE2zqXBjrl_YEpqyPVIIENYDuaa2v4f_t-ErgUQb3QN7jOKVF_QhOru3NQyWcZcBPEUlStTnayT1V0LK_2yjApvG9tLGgpZVGdsza9sbMexUQjRu3phFS1maVP3nV3Ee5D2mWWCzOwqjRvgOObzR4XxY8-9hIDI7iho3mNhpoiNubvH4s2ocmfImHmZqRBuLB-pNZ4dSAAnPIBDLpVWLmLLnxp5zIs2phfI7_GrHAz3LUj1ODVT2lyBjQYI_bNgpBHzfKhNfMBnHm4HpC_d7v7tD4X1cD_dOCBu3VQIs77u60sjOemot7g_F__P9GdrJQC60jZyWnW7IOE8V1Kg1ZtRISRBGavq3kj37H1qLHqEE6_R4tIgUvQiWb-9BoegnTKaX6_Cpq4L_OFGiegOr7EalAf2KcDm1AL15FEcHuYlHe5YeW-XGaHQdG7Ovt2owKAO1Kg; path=/; secure; samesite=lax; httponly
status: 302
x-correlation-id: d5a7b7fe-c5e9-4ab0-b6ae-0cb3ba891827

FORM-DATA
state=CfDJ8L16rcoVAjJMt5TusKivrFf2GEKRNyWN6Am9QMgqgxz17T9ThyCs-KUxjv1eUpwqojsNxikmLmsC5QXNBURb4gj8IzcQ1lYVPx4iP20oEk3bO2ZQgibTGRDlkBOAp6S7fSCUOPi1lwI1a5H9JjmewSY_4lpEftzbPKn9PCIbHuIkEcbh1UIDkptbanBb-ALt40tmtX8_UzgR7ZQJLLzzNp7POLvKSSXR8qugXU8CefSHCJGGNFM4GnNEVkRVW4oIhm2afXjlLJ13D90yaa8XzVPz_gHAiCgrCM4LVaDqTBfjHWZDiDAyDl66H_XGTaYjyNosKGStsZR3pF-23O0BQrmaXriT6hPTxKQnimMytT3DREAG0G6otxk_mBBMbQziepTeIqNWAMePCEOXZ5Ha367R343Kzyill8D_YNt82uTcFmf9xe-pAT7Mn9Q6P-p0Vsvs5WWN8dQyXposYpUThN-Ub6IoK4WtnNI2gYKBMR_S&code=REDACTED&scope=profile+openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile&authuser=0&prompt=none
```
```ExternalLoginModel.OnGetCallbackAsync``` gets called but then I get a ```null``` when calling ```GetExternalLoginInfoAsync ```  

I recall that some cookies get cleared during this exchange, but why would simply hosting my app in an iFrame result in different behavior?

**Comment by blowdart** [comment]
Because google messed with same site for privacy reasons. I assume this is in chrome? Is there any errors in the log files?

**Comment by ghstahl** [comment]
This is my beloved Edge.  I don't think its google, because I am getting a signin-google post back just fine.    Also, I have other OIDC idps ( "authority": "https://demo.identityserver.io/") as well as and azure-ad and they all give me the same problem.

I turned on "DEBUG"
```
[15:04:26 DBG][CorrelationId=26758291-1cd7-4da6-aeac-386b3fc6fea4] HandleChallenge with Location: https://accounts.google.com/o/oauth2/v2/auth?client_id=1096301616546-edbl612881t7rkpljp3qa3juminskulo.apps.googleusercontent.com&redirect_uri=https%3A%2F%2F127.0.0.1.xip.io%3A5001%2Fsignin-google&response_type=code&scope=openid%20profile&code_challenge=5s77lnzA02fd2klAHsocHQVriwQqx5Cue2B5zXZ71Wk&code_challenge_method=S256&response_mode=form_post&nonce=637342490664137218.NTRhNGNiNjgtZTJiNi00NTM1LTlhZWItYWI1OWRhNmVhNTBmMjFiZTMzMjktNzllMi00MjYyLTg5YzYtY2ZkNDRjYTgwMTMy&prompt=none&idp_code=DT&custom=%7B%22someCustomJson%22%3A%22hi%22%7D&state=CfDJ8L16rcoVAjJMt5TusKivrFfep58eYvSiI_uZvWmz_e7B_x_JaqATRg8ACowjLZ9TzlmDl97o7SnMDpkXXCm_fTmfWkEAH4ZX5Fh8Nui9lM-2KedNzUVpMzu96FEnxWz1_gXcD8Hf-0u-p6d_iOCB6_7Tc105z1_hILDQkzmwHBS7zRJkyX_U7-l66XVlnPNF9ylIxQuMnPoCWQf0CUOWrTlzY6EfCr7NgJPeAkxCVzPxASpgKTltG8_Yy88OrIICM1uxaTSgMgKZtQ31WA6cZ-Rfgk4J-lr1xqLvmRUuKNuGdpDi3BRCFdR120bvw-EyPSac_49utuDFxUmuslE_o3qr9crDiyCtl-JWWUc_RMSvSVOGCpGIv_Q1iDiSAz1Y8Er2_cGNkwliitNTTbKjf5HCjmWQwW_KNRapNSEYIm9aKdfilCX1SL3-42XBMVeVnh6Zc4h7Wt2Ic-fwwM-ed6mN2YurMjREd6jKqQC_MGWHW89Nm2RmdiFktzzlGInC1g&x-client-SKU=ID_NETSTANDARD2_0&x-client-ver=5.5.0.0; and Set-Cookie: .AspNetCore.OpenIdConnect.Nonce.CfDJ8L16rcoVAjJMt5TusKivrFdEyHxWYR7HjsZI_I8Pt_GzCcrX7kMhGKhvYrtEF1vM_oX_tNsDlpOSc3HSOfXDNgmB3fvScyg3OD8BlUf3Pc2VKOn0X16WaYl9oXkU_i5oZagdtINcw78FwZZ6bcvnMIAa1IzzWrq7bPkEJ9p3Db4nl_or6psnIjqj8_98_sFRqRayowAmaE24kQwWqmfF2XIAEEMhXx_kITxLwLpZXmEZb-Uvl_5y1Vm2WXtGW4lF3p0e3nCk40VndSEoO58x9OM=N; expires=Fri, 28 Aug 2020 22:19:26 GMT; path=/signin-google; secure; samesite=none; httponly,.AspNetCore.Correlation.google.49FznNh82pVsd9_BJDn7xQr8N0AU6danV0CknJwAhoU=N; expires=Fri, 28 Aug 2020 22:19:26 GMT; path=/signin-google; secure; samesite=none; httponly,.oidc.memoryCacheKey=%22a7e50bbd-3b1f-499d-87b9-0fa3b7ad1637%22; expires=Fri, 28 Aug 2020 23:04:26 GMT; path=/; samesite=none.
[15:04:26 INF][CorrelationId=26758291-1cd7-4da6-aeac-386b3fc6fea4] AuthenticationScheme: google was challenged.
[15:04:26 DBG][CorrelationId=26758291-1cd7-4da6-aeac-386b3fc6fea4] Cookie '.AspNetCore.Mvc.CookieTempDataProvider' suppressed due to consent policy.
[15:04:26 INF][CorrelationId=26758291-1cd7-4da6-aeac-386b3fc6fea4] Executed page /Account/ExternalLogin in 690.8256ms
[15:04:26 INF][CorrelationId=26758291-1cd7-4da6-aeac-386b3fc6fea4] Executed endpoint '/Account/ExternalLogin'
[15:04:26 DBG][CorrelationId=26758291-1cd7-4da6-aeac-386b3fc6fea4] Disposing the correlation context for this request
[15:04:26 DBG][CorrelationId=26758291-1cd7-4da6-aeac-386b3fc6fea4] Writing correlation ID response header X-Correlation-ID with value 26758291-1cd7-4da6-aeac-386b3fc6fea4
[15:04:26 INF][CorrelationId=26758291-1cd7-4da6-aeac-386b3fc6fea4] Request finished in 764.3989ms 302
[15:04:26 INF][CorrelationId=] Request starting HTTP/2 POST https://127.0.0.1.xip.io:5001/signin-google application/x-www-form-urlencoded 724
[15:04:26 DBG][CorrelationId=] POST requests are not supported
[15:04:26 DBG][CorrelationId=] Running correlation ID processing
[15:04:26 INF][CorrelationId=] No correlation ID was found in the request headers
[15:04:26 DBG][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] Generated a correlation ID 538a3bd4-a71b-4413-b9b8-f3e86da96892 using the CorrelationId.Providers.GuidCorrelationIdProvider provider
[15:04:26 DBG][CorrelationId=] Creating the correlation context for this request
[15:04:26 DBG][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] Correlation ID processing was completed with a final correlation ID 538a3bd4-a71b-4413-b9b8-f3e86da96892
[15:04:26 DBG][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] No candidates found for the request path '/signin-google'
[15:04:26 DBG][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] Request did not match any endpoints
[15:04:27 DBG][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] Connection id "0HM2BJP5N79LP", Request id "0HM2BJP5N79LP:00000013": started reading request body.
[15:04:27 DBG][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] Connection id "0HM2BJP5N79LP", Request id "0HM2BJP5N79LP:00000013": done reading request body.
[15:04:27 DBG][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] Updating configuration
[15:04:27 DBG][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] Redeeming code for tokens.
[15:04:27 INF][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] AuthenticationScheme: Identity.External signed in.
[15:04:27 DBG][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] Disposing the correlation context for this request
[15:04:27 DBG][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] Writing correlation ID response header X-Correlation-ID with value 538a3bd4-a71b-4413-b9b8-f3e86da96892
[15:04:27 INF][CorrelationId=538a3bd4-a71b-4413-b9b8-f3e86da96892] Request finished in 515.6951ms 302
[15:04:27 INF][CorrelationId=] Request starting HTTP/2 GET https://127.0.0.1.xip.io:5001/Identity/Account/ExternalLogin?returnUrl=%2FOIDCIFrameResult&handler=Callback
[15:04:27 DBG][CorrelationId=] The request path /Identity/Account/ExternalLogin does not match a supported file type
[15:04:27 DBG][CorrelationId=] Running correlation ID processing
[15:04:27 INF][CorrelationId=] No correlation ID was found in the request headers
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Generated a correlation ID b5d9eec2-fbe0-451f-b37e-c02fa99065e6 using the CorrelationId.Providers.GuidCorrelationIdProvider provider
[15:04:27 DBG][CorrelationId=] Creating the correlation context for this request
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Correlation ID processing was completed with a final correlation ID b5d9eec2-fbe0-451f-b37e-c02fa99065e6
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] 1 candidate(s) found for the request path '/Identity/Account/ExternalLogin'
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Endpoint 'Page: /Account/ExternalLogin' with route pattern 'Identity/Account/ExternalLogin' is valid for the request path '/Identity/Account/ExternalLogin'
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Request matched endpoint '/Account/ExternalLogin'
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] AuthenticationScheme: Identity.Application was not authenticated.
[15:04:27 INF][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Executing endpoint '/Account/ExternalLogin'
[15:04:27 INF][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Route matched with {page = "/Account/ExternalLogin", area = "Identity", action = "", controller = ""}. Executing page /Account/ExternalLogin
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Execution plan of authorization filters (in the following order): ["None"]
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Execution plan of resource filters (in the following order): ["Microsoft.AspNetCore.Mvc.ViewFeatures.Filters.SaveTempDataFilter"]
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Execution plan of action filters (in the following order): ["Microsoft.AspNetCore.Mvc.ModelBinding.UnsupportedContentTypeFilter (Order: -3000)", "InMemoryIdentityApp.SecurityHeadersAttribute (Order: 0)"]
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Execution plan of exception filters (in the following order): ["None"]
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Execution plan of result filters (in the following order): ["Microsoft.AspNetCore.Mvc.ViewFeatures.Filters.SaveTempDataFilter", "InMemoryIdentityApp.SecurityHeadersAttribute (Order: 0)"]
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Executing page model factory for page AspNetCore.Areas_Identity_Pages_Account_ExternalLogin (InMemoryIdentityApp.Views)
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Executed page model factory for page AspNetCore.Areas_Identity_Pages_Account_ExternalLogin (InMemoryIdentityApp.Views)
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Attempting to bind property 'InMemoryIdentityApp.Areas.Identity.Pages.Account.ExternalLoginModel.Input' of type 'InMemoryIdentityApp.Areas.Identity.Pages.Account.ExternalLoginModel+InputModel' ...
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Skipped binding property 'InMemoryIdentityApp.Areas.Identity.Pages.Account.ExternalLoginModel.Input' since its binding information disallowed it for the current request.
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Attempting to bind parameter 'returnUrl' of type 'System.String' ...
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Attempting to bind parameter 'returnUrl' of type 'System.String' using the name 'returnUrl' in request data ...
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Done attempting to bind parameter 'returnUrl' of type 'System.String'.
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Done attempting to bind parameter 'returnUrl' of type 'System.String'.
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Attempting to validate the bound parameter 'returnUrl' of type 'System.String' ...
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Done attempting to validate the bound parameter 'returnUrl' of type 'System.String'.
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Attempting to bind parameter 'remoteError' of type 'System.String' ...
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Could not find a value in the request with name '' for binding parameter 'remoteError' of type 'System.String'.
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Done attempting to bind parameter 'remoteError' of type 'System.String'.
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Done attempting to bind parameter 'remoteError' of type 'System.String'.
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Attempting to validate the bound parameter 'remoteError' of type 'System.String' ...
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Done attempting to validate the bound parameter 'remoteError' of type 'System.String'.
[15:04:27 DBG][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] The temp data cookie .AspNetCore.Mvc.CookieTempDataProvider was not found.
[15:04:27 INF][CorrelationId=b5d9eec2-fbe0-451f-b37e-c02fa99065e6] Executing handler method InMemoryIdentityApp.Areas.Identity.Pages.Account.ExternalLoginModel.OnGetCallbackAsync - ModelState is Valid
```

**Comment by Tratcher** [comment]
Can you show the raw request trace for /Identity/Account/ExternalLogin? Did it even send the cookie? It's an awfully big cookie.

**Comment by ghstahl** [comment]
[test app](https://github.com/fluffy-bunny/asp.net-inmemory-identity-template).
[har files](https://github.com/dotnet/aspnetcore/files/5145197/asp.net-inmemory-identity-template.zip)

What I can tell between the 2
```https://localhost:7001/plugin```
vs
```https://localhost:7001/IFrameHost/Plugin``` (hosts https://127.0.0.1.xip.io:7001/plugin in an iframe)

#  Working no IFrame
The ```https://localhost:7001/plugin``` version drops a big ```set-cookie: Identity.External``` cookie here ```https://localhost:7001/signin-google```  

when ```https://localhost:7001/Identity/Account/ExternalLogin?returnUrl=%2FOIDCIFrameResult&handler=Callback``` is called there are some big cookies in the request.

# Not Working - iFrame
```https://127.0.0.1.xip.io:7001:/plugin``` hosted in an iFrame version drops a big ```set-cookie: Identity.External``` cookie here ```https://127.0.0.1.xip.io:7001/signin-google```   

when ```https://127.0.0.1.xip.io:7001/Identity/Account/ExternalLogin?returnUrl=%2FOIDCIFrameResult&handler=Callback``` is called the only cookie in the request is ```cookie: .oidc.memoryCacheKey=%226900eaca-fa6c-43df-b699-0b357ca203d7%22```  

I am loosed all cookies between ```signin-google``` and ```ExternalLogin```.

**Comment by ghstahl** [comment]
@Tratcher I was stepping the code using symbols and I put a breakpoint in [ChunkingCookieManager](https://github.com/dotnet/aspnetcore/blob/v3.1.7/src/Shared/ChunkingCookieManager/ChunkingCookieManager.cs) ```AppendResponseCookie```.

I changed the ```Samesite``` from ```LAX``` to ```NONE``` and that seemed to now let cookies to be dropped whilst in the IFrame.  I noticed that my cookie ```.oidc.memoryCacheKey``` sas the only cookie in the IFrame in the failure case.
I use the following Cookie Options.
```
CookieOptions option = new CookieOptions()
            {
                IsEssential = true,
                SameSite = SameSiteMode.None,
                Secure = true
            };
```

What is strange is that I tried to override the default ChunkingCookieManager with mine and blindly changes the Samesite to ```none``` on all ```AddResponseCookie```
```
 services.ConfigureApplicationCookie(options =>
  {
     options.CookieManager = new ChunkingCookieManager2();
...
}
```
My ChunkingCookieManager2's ```AppendResponseCookie``` gets hit sometimes but not all and it still fails.  

I wish I could get a buildable v3.1.7 aspnetcore but the Nuget.config is referencing internal repositories.

**Comment by Tratcher** [comment]
`.oidc.memoryCacheKey=%226900eaca-fa6c-43df-b699-0b357ca203d7%22; expires=Sat, 29 Aug 2020 13:14:39 GMT; path=/; samesite=none` is missing the `secure` attribute. That's required when using SameSite=None. However, that cookie seems to be returned anyways.

The Identity.External also probobly needs SameSite=None when working in an IFrame. You'd set that via ConfigureExternalCookie. You shouldn't need a custom cookie manager, you can set it directly on the options.Cookie.SameSite.

**Comment by ghstahl** [comment]
I made sure my ```.oidc.memoryCacheKey``` cookies is secure now.
```
 services.ConfigureExternalCookie(config =>
{
  config.Cookie.SameSite = SameSiteMode.None;
});
services.ConfigureApplicationCookie(options =>
{
   ....
}
```
The above did the trick for ```Identity.External```, however now downstream is having what appears to be the same issues.
My ```session cookie is lax``` and not getting dropped and my ```application cookie is lax``` isn't getting dropped either.  

The orginal problem is gone, ```GetExternalLoginInfoAsync ``` is now returning good stuff.

Downstream fixes (or are they?)
```
 services.AddSession(options =>
                {
                    options.Cookie.IsEssential = true;
                    options.Cookie.Name = $"{Configuration["applicationName"]}.Session";
                // Set a short timeout for easy testing.
                    options.IdleTimeout = TimeSpan.FromSeconds(3600);
                    options.Cookie.HttpOnly = true;
                    options.Cookie.SameSite = SameSiteMode.None;
                    options.Cookie.SecurePolicy = CookieSecurePolicy.Always;
                });
```
```
services.ConfigureApplicationCookie(options =>
 {
                    options.Cookie.SameSite = SameSiteMode.None;
                    options.Cookie.SecurePolicy = CookieSecurePolicy.Always;
}
```

Did I just do something **_bad_** with setting all my cookies ```samesite``` to ```none``` ?

 
BTW: Works on Google Chrome and Microsoft Edge.
For chrome I had to make it ESSENTIAL+NONE+SECURE, Edge just needed ESSENTIAL.  
Edge works with ESSENTIAL+NONE+SECURE as well, so that's good.

**Comment by Tratcher** [comment]
Edge will start enforcing the same SameSite restrictions sooner or later.

> Did I just do something **_bad_** with setting all my cookies `samesite` to `none` ?

You did something that Chrome is trying to convince people to stop doing. It makes you slightly more vulnerable to CSRF attacks, or at least just as vulnerable as you were before chrome started enforcing this.

Your choices are to give up on IFrames, or to live with SameSite=none.

**Comment by ghstahl** [comment]
My hands are tied as the host can only bring me in via IFrame at the moment.  I proved I can dynamically load any javascript SPA from anywhere thus still letting external teams be independent.  They only need to make sure their SPA's work good client side.  

I tried to load Blazor WASM's from anywhere (Blazor WASM is on some random CDN)  but at the moment they don't support that.    Thanks for the clues.
